File: tests/fixtures/gcs-ts/basic/storage.ts
import { Storage } from '@google-cloud/storage';

const storage = new Storage();

export async function fetchObject() {
  // GET → gs://static-assets/logo.png
  return storage.bucket('static-assets').file('logo.png').download();
}

export async function uploadObject() {
  // PUT → gs://user-uploads/inbox/new.txt
  return storage.bucket('user-uploads').file('inbox/new.txt').save(Buffer.from('hi'));
}

export async function deleteObject() {
  // DELETE → gs://archive/2026/snapshot.tar
  return storage.bucket('archive').file('2026/snapshot.tar').delete();
}

export async function headObject() {
  // GET (exists) → gs://static-assets/logo.png
  return storage.bucket('static-assets').file('logo.png').exists();
}

export async function getMetadataObject() {
  // GET → gs://static-assets/logo.png
  return storage.bucket('static-assets').file('logo.png').getMetadata();
}

export async function uploadFromPath() {
  // PUT → gs://user-uploads/ (bucket scope)
  return storage.bucket('user-uploads').upload('/local/path.txt');
}

export async function listFilesInBucket() {
  // GET → gs://static-assets/ (bucket scope)
  return storage.bucket('static-assets').getFiles();
}

export async function deleteBucket() {
  // DELETE → gs://temp-bucket/ (no .file() in chain)
  return storage.bucket('temp-bucket').delete();
}

export async function copyObject() {
  // POST → gs://archive/old.tar
  return storage
    .bucket('archive')
    .file('old.tar')
    .copy(storage.bucket('archive').file('new.tar'));
}

export async function setMetadataObject() {
  // PATCH → gs://configs/app.json
  return storage.bucket('configs').file('app.json').setMetadata({ contentType: 'application/json' });
}

export async function dynamicBucket(name: string) {
  // GET (dynamic) → null
  return storage.bucket(name).file('logo.png').download();
}

export async function dynamicKey(key: string) {
  // GET (bucket exact, key dynamic) → gs://static-assets/ (dynamic confidence)
  return storage.bucket('static-assets').file(key).download();
}

export async function getSignedDownloadUrl() {
  // GET → gs://static-assets/logo.png
  return storage.bucket('static-assets').file('logo.png').getSignedUrl({ action: 'read', expires: Date.now() + 1000 });
}

File: tests/fixtures/gcs-go/basic/storage.go
package main

import (
	"context"
	"io"

	"cloud.google.com/go/storage"
)

func fetchObject(ctx context.Context, client *storage.Client) (io.ReadCloser, error) {
	// GET → gs://static-assets/logo.png
	return client.Bucket("static-assets").Object("logo.png").NewReader(ctx)
}

func uploadObject(ctx context.Context, client *storage.Client) *storage.Writer {
	// PUT → gs://user-uploads/inbox/new.txt
	return client.Bucket("user-uploads").Object("inbox/new.txt").NewWriter(ctx)
}

func deleteObject(ctx context.Context, client *storage.Client) error {
	// DELETE → gs://archive/2026/snapshot.tar
	return client.Bucket("archive").Object("2026/snapshot.tar").Delete(ctx)
}

func headObject(ctx context.Context, client *storage.Client) (*storage.ObjectAttrs, error) {
	// GET (Attrs) → gs://static-assets/logo.png
	return client.Bucket("static-assets").Object("logo.png").Attrs(ctx)
}

func updateObject(ctx context.Context, client *storage.Client) (*storage.ObjectAttrs, error) {
	// PATCH → gs://configs/app.json
	return client.Bucket("configs").Object("app.json").Update(ctx, storage.ObjectAttrsToUpdate{})
}

func listFilesInBucket(ctx context.Context, client *storage.Client) *storage.ObjectIterator {
	// GET → gs://static-assets/ (bucket scope)
	return client.Bucket("static-assets").Objects(ctx, nil)
}

func deleteBucket(ctx context.Context, client *storage.Client) error {
	// DELETE → gs://temp-bucket/ (bucket scope; no .Object() in chain)
	return client.Bucket("temp-bucket").Delete(ctx)
}

func copierStart(ctx context.Context, client *storage.Client, src *storage.ObjectHandle) {
	// POST → gs://archive/dest.tar (CopierFrom)
	_ = client.Bucket("archive").Object("dest.tar").CopierFrom(src)
}

func dynamicBucket(ctx context.Context, client *storage.Client, name string) (io.ReadCloser, error) {
	// GET (dynamic bucket) → null URL
	return client.Bucket(name).Object("logo.png").NewReader(ctx)
}

func dynamicKey(ctx context.Context, client *storage.Client, key string) (io.ReadCloser, error) {
	// GET (literal bucket, dynamic key) → gs://static-assets/ (dynamic confidence)
	return client.Bucket("static-assets").Object(key).NewReader(ctx)
}

func rawStringBucket(ctx context.Context, client *storage.Client) (io.ReadCloser, error) {
	// GET → gs://raw-bucket/raw-key (raw string literals)
	return client.Bucket(`raw-bucket`).Object(`raw-key`).NewReader(ctx)
}
